bindings/cpp/src/lib.rs

```diff
@@ -56,7 +83,120 @@ impl Operator {
         Ok(self.0.read(path)?)
     }
 
-    fn write(&self, path: &str, bs: &[u8]) -> Result<()> {
-        Ok(self.0.write(path, bs.to_owned())?)
+    // To avoid copying the bytes, we use &'static [u8] here.
+    //
+    // Safety: The bytes created from bs will be dropped after the function call.
+    // So it's safe to declare its lifetime as 'static.
+    fn write(&self, path: &str, bs: &'static [u8]) -> Result<()> {
+        Ok(self.0.write(path, bs)?)
+    }
+
+    fn is_exist(&self, path: &str) -> Result<bool> {
+        Ok(self.0.is_exist(path)?)
+    }
+
+    fn create_dir(&self, path: &str) -> Result<()> {
+        Ok(self.0.create_dir(path)?)
+    }
+
+    fn copy(&self, src: &str, dst: &str) -> Result<()> {
+        Ok(self.0.copy(src, dst)?)
+    }
+
+    fn rename(&self, src: &str, dst: &str) -> Result<()> {
+        Ok(self.0.rename(src, dst)?)
+    }
+
+    // We can't name it to delete because it's a keyword in C++
+    fn remove(&self, path: &str) -> Result<()> {
+        Ok(self.0.delete(path)?)
+    }
+
+    fn stat(&self, path: &str) -> Result<ffi::Metadata> {
+        Ok(self.0.stat(path)?.into())
+    }
+
+    fn list(&self, path: &str) -> Result<Vec<ffi::Entry>> {
+        Ok(self.0.list(path)?.into_iter().map(Into::into).collect())
+    }
+}
+
+impl From<od::Metadata> for ffi::Metadata {
+    fn from(meta: od::Metadata) -> Self {
+        let mut tag = 0u8;
+
+        match meta.mode() {
+            od::EntryMode::FILE => tag |= 0b0000_0001,
+            od::EntryMode::DIR => tag |= 0b0000_0010,
+            _ => {}
+        }
+
+        let content_length = meta.content_length();
+
+        let cache_control = match meta.cache_control() {
+            Some(v) => {
+                tag |= 0b0000_0100;
+                v.to_owned()
+            }
+            None => String::new(),
```

Comment:
We can make `ffi::Metadata` store `&str` rather than `String` to avoid unnecessary construct about `String`. But this need some extra work to deal with lifecycle and ownership. Let me have a try.

Reply:
> We can make `ffi::Metadata` store `&str` rather than `String` to avoid unnecessary construct about `String`.

We don't need to. We can copy and move here directly.